mesonbuild/mtest.py

```diff
@@ -1859,17 +1860,34 @@ def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestS
         run all tests with that name across all subprojects, which is
         identical to "meson test foo1"
         '''
+        patterns = {}
         for arg in self.options.args:
             if ':' in arg:
                 subproj, name = arg.split(':', maxsplit=1)
+                if name == '':
+                    name = '*'
+                if subproj == '':  # in case arg was ':'
+                    subproj = '*'
             else:
-                subproj, name = '', arg
-            for t in tests:
-                if subproj and t.project_name != subproj:
-                    continue
-                if name and t.name != name:
-                    continue
-                yield t
+                subproj, name = '*', arg
+            patterns[(subproj, name)] = False
+
+        for t in tests:
+            for subprojp, namep in list(patterns):
+                if fnmatch(t.project_name, subprojp) and fnmatch(t.name, namep):
+                    patterns[(subprojp, namep)] = True
+                    yield t
+                    break
+
+        for (subprojp, namep), was_used in patterns.items():
+            if not was_used:
+                arg = f'{subproj}:{namep}'
+                for t in tests:
+                    if fnmatch(t.project_name, subprojp) and fnmatch(t.name, namep):
+                        mlog.warning(f'{arg} test name is redundant and was not used')
```

Comment:
I'm having a hard time reading this... when does it actually trigger?

Reply:
Okay... I think what this is doing is fnmatch'ing its way around all the tests and checking if an unmatched argument was skipped because we broke out of pattern matching as soon as a test was eligible for a single reason. Hence even though it matches in both cases, we never set was_used.

So we don't know whether it is an erroring match, or just a redundant one.

Can you please add a comment about this?